plutus-executables/executables/uplc/Main.hs

```diff
@@ -272,16 +281,62 @@ runOptimisations (OptimiseOptions inp ifmt outp ofmt mode cert) = do
-                (Right before', Right after')             -> (stage, (AgdaFFI.conv (void before'), AgdaFFI.conv (void after')))
-                (Left (err :: UPLC.FreeVariableError), _) -> error $ show err
-                (_, Left (err :: UPLC.FreeVariableError)) -> error $ show err
-          rawAgdaTrace = reverse $ processAgdaAST <$> simplTrace
-      Agda.runCertifier (T.pack certName) rawAgdaTrace
-    runCertifier Nothing _ = pure ()
+
+---------------- Agda certifier ----------------
+
+-- | Run the Agda certifier on the simplification trace, if requested
+runCertifier
+  :: Maybe String
+  -- ^ Should we run the Agda certifier? If so, what should the certificate file be called?
+  -> SimplifierTrace UPLC.Name UPLC.DefaultUni UPLC.DefaultFun a
+  -- ^ The trace produced by the simplification process
+  -> IO ()
+runCertifier (Just certName) (SimplifierTrace simplTrace) = do
+  let processAgdaAST Simplification {beforeAST, stage, afterAST} =
+          case (UPLC.deBruijnTerm beforeAST, UPLC.deBruijnTerm afterAST) of
+            (Right before', Right after')             ->
+              (stage, (AgdaFFI.conv (void before'), AgdaFFI.conv (void after')))
+            (Left (err :: UPLC.FreeVariableError), _) -> error $ show err
+            (_, Left (err :: UPLC.FreeVariableError)) -> error $ show err
+      rawAgdaTrace = reverse $ processAgdaAST <$> simplTrace
+  runAgda certName rawAgdaTrace
+runCertifier Nothing _ = pure ()
+
+-- | Run the Agda compiler on the metatheory and evaluate the 'runCertifier' function
+-- on the given trace.
+runAgda
+  :: String
+  -- ^ The name of the certificate file to write
+  -> [(SimplifierStage, (AgdaFFI.UTerm, AgdaFFI.UTerm))]
+  -- ^ The trace produced by the simplification process
+  -> IO ()
+runAgda certName rawTrace = do
+  let program = "runCertifier (" ++ agdaUnparse rawTrace ++ ")"
+  (parseTraceResult, _) <- HAgda.Parser.runPMIO $ HAgda.Parser.parse HAgda.Parser.exprParser program
+  let parsedTrace =
+        case parseTraceResult of
+          Right (res, _) -> res
+          Left err       -> error $ show err
+  stdlibPath <- getEnv "AGDA_STDLIB_SRC"
+  metatheoryPath <- getEnv "PLUTUS_METHATHEORY_SRC"
+  let inputFile = HAgda.File.AbsolutePath (Text.pack $ metatheoryPath </> "Certifier.agda")
+  runTCMPrettyErrors $ do
+    let opts =
+          defaultOptions
+            { optIncludePaths =
+                [ metatheoryPath
+                , stdlibPath
+                ]
+            }
+    setCommandLineOptions opts
+    result <- HAgda.Imp.typeCheckMain HAgda.Imp.TypeCheck =<< HAgda.Imp.parseSource (HAgda.File.SourceFile inputFile)
+    let interface = crInterface result
+        insideScope = iInsideScope interface
+    setScope insideScope
+    internalisedTrace <- toAbstract parsedTrace
+    decisionProcedureResult <- evalInCurrent DefaultCompute internalisedTrace
+    final <- prettyTCM decisionProcedureResult
+    liftIO $ writeFile (certName ++ ".agda") (show final)
```

Comment:
Could you show some small example file?

Reply:
I can't attach one here, see Slack: https://input-output-rnd.slack.com/archives/G01EE4NQ9U0/p1729780865708759